[Lee Cheng Da] iP

chengda300 opened the pull request:
# Dude
> It sounds like Duke, but my name is actually Dude. -👦 

This bot *can* manage a list of tasks by:
- Specifying deadlines
- Marking a task's completion status
- Filtering task names within the list of tasks

To **use** this bot:
1. Load previous run's data if any
2. Add tasks to list
3. Mark tasks as completed, or delete tasks

This bot has supported the lives of:
- [x] 1 person
- [ ] 10 people
- [ ] 100 people
- [ ] 1,000 people
- [ ] 10,000 people
- [ ] 100,000 people
- [ ] 1,000,000 people

This is how this bot is created using the `Duke` constructor method.
```java
public Duke() {
    this.ui = new Ui();
    this.storage = new Storage(this.filePath);
    this.tasks = new TaskList();
    this.parser = new Parser();
}
```

**This bot is [open source](https://github.com/chengda300/ip) too!**

chengda300 committed:
Level-1

chengda300 committed:
Level-2

chengda300 committed:
Level-3

chengda300 committed:
Level-1

chengda300 committed:
Level-2

chengda300 committed:
Level-3

chengda300 committed:
Level-4

chengda300 committed:
A-TextUiTesting

chengda300 committed:
Level-5

chengda300 committed:
Level-6

chengda300 committed:
Fully consolidate control flow from the main method to the respective methods handling the functions

chengda300 committed:
Add level 7

chengda300 committed:
Add Level-8

chengda300 committed:
Merge branch 'branch-Level-7'

chengda300 committed:
Merge branch 'branch-Level-8'

chengda300 committed:
Make code more OOP

chengda300 committed:
Consolidate classes into a package

chengda300 committed:
Add gradle

chengda300 committed:
Add junit tests

chengda300 committed:
Release v0.1

chengda300 committed:
Add documentation

chengda300 committed:
Revert "Add documentation"

This reverts commit 7e752882dd0eade1f690b271dfe43048e500766c.

chengda300 committed:
Add documentation

chengda300 committed:
Add a new feature called find

chengda300 committed:
Merge branch 'branch-A-JavaDoc'

chengda300 committed:
Merge branch 'branch-Level-9'

chengda300 committed:
Add checkstyle

chengda300 committed:
Add Level-10 and reformat certain parts of the code

chengda300 committed:
Separate saving process from individual parsing methods

yiyefyy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,164 @@
+            }
+        } else if (input.startsWith(this.find + this.space)) {
+            String s = input.substring((this.find + this.space).length());
+            if (s.isBlank()) {
```
I think you can avoid deep nesting by using switch.

chengda300 replied:
Switch statements with the current code structure would be too complicated.
Instead, I will shift the bulk of the processes to private methods of the parser class and have the parser call those methods instead. Apparently, the code was not following the single level of abstraction principle which resulted in a very long parse method.

yiyefyy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,175 @@
+ * Represents the Duke bot.
+ */
+public class Duke {
+    private static boolean isLoading = true;
```
Nice naming!

chengda300 replied:
Thank you!

yiyefyy commented on src/test/java/TaskTest.java:
```diff
@@ -0,0 +1,25 @@
+    Todo t = new Todo("123");
+
+    @Test
+    void testGetStatusIcon() {
```
you can use underscores to name the test methods: eg test_Find_Success.

chengda300 replied:
Ok, I will adjust the names

yiyefyy reviewed (commented):
I think your code is quite good and mostly followed the coding standards.

nqt230 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,164 @@
+ * Represents a parser to parse inputs into actions to be taken.
+ */
+public class Parser {
+    private String bye = "bye";
```
I like how you define the command strings as a class field. You can also consider making these constant, since the commands are not likely to change over the course of running the program (unless you support command remapping).

chengda300 replied:
Yes, it should have been private static final instead, I will change them.
Thank you!

nqt230 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,164 @@
+     * @param isLoading Boolean to indicate if the Duke bot is loading.
+     */
+    public void parse(Duke d, String input, boolean isLoading) {
+        if (input.equals(this.bye)) {
```
Since there are multiple if-else statements to check the input against the possible commands, would a switch statement be better here for readability?

chengda300 replied:
Switch statements with the current code structure would be too complicated.
Instead, I will shift the bulk of the processes to private methods of the parser class and have the parser call those methods instead. Apparently, the code was not following the single level of abstraction principle which resulted in a very long parse method.

nqt230 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,164 @@
+                    System.out.println("The task is empty, what do you really mean?");
+                } else if (deadline.isBlank()) {
+                    System.out.println("The deadline is empty, do you mean it has no deadline?");
+                    System.out.println("If it is, please add it as a todo instead.");
```
I like the user feedback you give here, I can see how these prompts would be very helpful for a first time user.

nqt230 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,63 @@
+        File f = new File(filePath);
+        Scanner s = new Scanner(f);
+        while (s.hasNext()) {
+            d.parse(s.nextLine());
```
It seems your approach to loading back the data is to save the entire command strings and parse them when you want to load the data in, reconstructing the tasks this way. I think this could potentially result in save file bloating if there are multiple repeated delete commands, although this is not a big issue given the scale of the project. Would it be better to convert Tasks into some kind of save string format that you can use to reconstruct the Task?

chengda300 replied:
Yes, it is a lazy way of storing and loading data for small scale projects.
For large scale projects it should undo the action as if the previous action did not happen at all to keep the list as short as possible.

This lazy way allows autosave in case the program crashes halfway. It is possible to add in an optimiser function during loading to debloat the text file.

nqt230 reviewed (commented):
Looks good :+1:
Just some minor concerns regarding the approach to saving of data and parsing of input, but I think overall the code is good.

chengda300 committed:
Fix resizing GUI bugs

totsukatomofumi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,287 @@
+     * @return Associated message from Duke.
+     */
+    public String addDeadline(String s, String d) {
+        return this.tasks.addDeadline(s, d);
```
I like the use of overloading methods!

chengda300 replied:
Thank you!

totsukatomofumi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,287 @@
+     * @param s String to request for tasks with the matching string.
+     * @return List of tasks with matching string.
+     */
+    public String find(String s) {
```
Would it be better if the method naming was a bit more descriptive?

chengda300 replied:
Yes, I will change it to findTask, thank you

totsukatomofumi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,287 @@
+        stage.setScene(scene);
+        stage.show();
+
+        //Step 2. Formatting the window to look as expected
```
I like the comments that help explain the underlying code.

chengda300 replied:
Actually that was in the GUI guide.

totsukatomofumi commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,51 @@
+ * Represents an event. An event has a description and a time that it will happen.
+ */
+public class Event extends Task {
+    private String at;
```
Will it be better to give this a slightly more descriptive name?

chengda300 replied:
Yes it will be, it should be changed to event time I guess.

totsukatomofumi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,164 @@
+ * Represents a parser to parse inputs into actions to be taken.
+ */
+public class Parser {
+    private static final String BYE = "bye";
+    private static final String LIST = "list";
+    private static final String MARK = "mark";
+    private static final String UNMARK = "unmark";
+    private static final String TODO = "todo";
+    private static final String DEADLINE = "deadline";
+    private static final String DEADLINE_BY = "/by";
+    private static final String EVENT = "event";
+    private static final String EVENT_AT = "/at";
+    private static final String DELETE = "delete";
+    private static final String SPACE = " ";
+    private static final String FIND = "find";
+    private static final String ERROR_MESSAGE = "Sorry, I cannot understand what you exactly mean.\n"
```
Very good use constants for values that may be used very often in the code.

totsukatomofumi reviewed (commented):
LGTM! Overall, I am impressed with the logic of code you have made. There are some variable names that may be improved, but I am just going by the book. Good job!

chengda300 committed:
Fix markTask, unmarkTask and deleteTask deleting the wrong task.

Previously, the methods take in the input as the index.
It is changed to take in the input as the numbered task in the list,
then it will internally adjust to the index to delete.

For example, to delete the first task in the list,
in the previous version it would be delete 0, now it will be delete 1.

chengda300 committed:
Revert "Fix markTask, unmarkTask and deleteTask deleting the wrong task."

This reverts commit 732c551f39daeae84693877c72725db4620c4fad.

chengda300 committed:
Revert "Revert "Fix markTask, unmarkTask and deleteTask deleting the wrong task.""

This reverts commit 3f4b0959e09d8dde9219925cc48c49131d099035.

chengda300 committed:
Improve code quality

Previously, certain methods that perform the same task have different
parameter names. As such, parameter names are changed to be more
consistent in general. For example, delete() has int i as the method
parameter, the same way as markTask() and unmarkTask(), while it has
int index as the method parameter previously.

This would improve consistency across the code base.

chengda300 committed:
Merge pull request #1 from chengda300/branch-Assertions

Branch assertions

chengda300 committed:
Merge branch 'master' of https://github.com/chengda300/ip

chengda300 committed:
Merge branch 'master' into branch-CodeQuality

chengda300 committed:
Merge pull request #2 from chengda300/branch-CodeQuality

Improve code quality

chengda300 committed:
Merge branch 'master' of https://github.com/chengda300/ip

chengda300 committed:
Merge branch 'branch-CodeQuality' of https://github.com/chengda300/ip into branch-CodeQuality

chengda300 committed:
Add Continuous Integration

chengda300 committed:
Create gradle.yml

chengda300 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	.github/workflows/gradle.yml

chengda300 committed:
Add ability to mark, unmark and delete multiple tasks, and also fix
an error of mark and unmark commands saving incorrectly

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Delete I/O redirection tests for CI

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Update execute permission for gradlew command

chengda300 committed:
Attempt to fix CI failures

chengda300 committed:
Fix some checkstyle errors

chengda300 committed:
Fix some checkstyle errors

chengda300 committed:
Prepare to release

chengda300 committed:
Fix minor issues with imports and user guide

chengda300 committed:
Fix user guide errors

chengda300 committed:
Fix Java compiler language level error
